assembly/src/parsers/u32_ops.rs

```diff
@@ -469,20 +494,31 @@ pub fn parse_u32shl(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
 /// be shifted is asserted to be a 32-bit value.
 ///
 /// VM cycles per mode:
-/// - u32shr: 47 cycles
-/// - u32shr.b: 4 cycles
-/// - u32shr.unsafe: 44 cycles
-pub fn parse_u32shr(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+/// - u32checked_shr: 47 cycles
+/// - u32checked_shr.b: 4 cycles
+/// - u32unchecked_shr: 40 cycles
+/// - u32unchecked_shr.b: 3 cycles
+/// - u32overflowing_shr: 44 cycles
+/// - u32overflowing_shr.b: 3 cycles
+pub fn parse_u32shr(
+    span_ops: &mut Vec<Operation>,
+    op: &Token,
+    op_mode: U32OpMode,
+) -> Result<(), AssemblyError> {
     let drop_remainder = match op.num_parts() {
         0 => return Err(AssemblyError::invalid_op(op)),
-        1 => {
-            // Assume the dynamic shift value b is on top of the stack.
-            aggregate_power_2(span_ops, false);
-            span_ops.push(Operation::U32assert2);
-            true
-        }
-        2 => match op.parts()[1] {
-            "unsafe" => {
+        1 => match op_mode {
+            U32OpMode::Checked => {
+                // Assume the dynamic shift value b is on top of the stack.
+                aggregate_power_2(span_ops, false);
+                span_ops.push(Operation::U32assert2);
```

Comment:
I'm not sure I understood what you meant. Now, during the execution of the `aggregate_power_2()` function in safe mode, we check whether the parameter is `u32` and whether it is less than 64. After executing the function, we check that the number already raised to the power and the original number to which the shift will be applied belong to `u32`.
In this case, after executing the `aggregate_power_2()` function, how can we check whether the original number on the stack was less than 64? After all, we will already be working with the exponentiated number.
It is not clear to me yet how we can do with one check, because in the first case it is needed to make sure that `U32div` will work correctly, and in the second case we check the final, resulting values

Reply:
OK - let's skip it.